### DISCUSSION TITLE
Batch loading a relationship on an already-loaded result set

### DISCUSSION THREAD

**Discussion by dkbarn** [sqlalchemy/sqlalchemy#10908] [Usage Questions]
What is the best way to handle the following scenario, where a collection of results have already been fetched from the db, and now I would like to query (ideally via a single SELECT IN statement) a relationship across all of them?

```python

results = session.execute(stmt).scalars().all()

filtered_results = []
for result in results:
    if condition:
        filtered_results.append(result)

# TODO: somehow trigger batch-loading of the relationship result.related_object for everything in filtered_results

for result in filtered_results:
    print(result.related_object)
```

The issue here is that eager loading `result.related_object` during the initial query would over-fetch, since they are only being accessed on a subset of all results. However, treating them as a lazy-loaded relationship is also a waste because it fails to batch them together into a single SELECT IN statement, which would be the case with eager loading. Finally, formulating the SELECT IN statement myself to fetch all the `related_object` entities as a separate query is also not ideal, because they would be returned as a disconnected result set, and not accessible via the `result.related_objects` attribute.

Does SQLAlchemy provide any mechanism for dealing with this scenario efficiently?

**Comment by dkbarn** [comment]
The best I've been able to come up with involves running separate query with `populate_existing=True`:

```python

results = session.execute(stmt).scalars().all()

filtered_results = []
for result in results:
    if condition:
        filtered_results.append(result)

# trigger batch-loading of the relationship result.related_object for everything in filtered_results
stmt2 = (
    select(Object)
        .filter(Object.id.in_(r.id for r in filtered_results))
        .options(
            load_only(Object.id),
            selectinload(Object.related_object)
        )
        .execution_options(populate_existing=True)
)
session.execute(stmt2).scalars().all()

for result in filtered_results:
    print(result.related_object)

```
This seems quite verbose and clunky though. Also, the second statement execution involves dispatching two SELECT statements: one to get the IDs of the Object entities, and a second to get the related_object entities, whereas only the second should be needed.

**Comment by zzzeek** [comment]
well this is obviously normally the "selectinload" pattern but here you are wanting to apply in-python filtering to the result before doing the selectinload operation.

so you are on the right track running a query but take it all the way, run the exact SELECT statement for the related objects you want directly, no two-query requirement:

```py
stmt2 = (
    select(RelatedObject)
        .filter(RelatedObject.foreign_key_id.in_(r.id for r in filtered_results)).order_by(RelatedObject.foreign_key_id)
)
related_objects = session.scalars(stmt2).all()

filtered_result_by_id = {r.id: r for r in filtered_results}
for fkid, group in in itertools.groupby(related_objects, lambda r: r.foreign_key_id):
    attributes.set_committed_value(filtered_result_by_id[fkid], "related_objects", list(group))
```

this is a paraphrase of the recipe at https://github.com/sqlalchemy/sqlalchemy/wiki/DisjointEagerLoading

**Reply by dkbarn** [reply]
Interesting.. `attributes.set_committed_value` looks like the key I was missing. Thanks for pointing me to that wiki recipe, looks like exactly what I was after!

Would it be a reasonable feature request to ask for a `session.refresh_all` method, which could perform some of this magic under the hood? It would be super handy if I could simply do:
```python
session.refresh_all(filtered_results, ["related_objects"])
```
Given the 2.0.4 change in behavior described [here](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.refresh) in which explicitly-named lazy-loaded relationships will be loaded by `session.refresh`, having this same functionality work across a collection of objects as efficiently as this disjoint eager loading recipe could be really useful -- with the obvious caveat that if attribute names are explicitly listed, the objects in the collection must all be of the same type.

**Reply by zzzeek** [reply]
99% of what refresh() does can be done for multiple objects using the populate_existing feature with a regular query.  the only thing it can't do is omit the primary query.   Omitting the primary query is only appropriate in that small set of cases where one only wants to load related objects and no local columns - loading of relationships and not columns is already not what refresh() is oriented towards since it was not originally designed for that.  refresh() does not even let you choose specific eager loader styles for attributes, so this does not seem to be the right place to build more of that pattern.

this is all to say that adding a refresh_all() further complicates an already too-complicated API, and also the current architecture of refresh() is not well suited so it would be a heavy lift that would require refactoring the internals a lot.

when we get to that place, where something would be a lot of work to do, and there's a lot that seems to not really work yet that would also have to be thought out, we look at the feature as is this something for which lots of people really want, all the time. Examples of features that were very involved but people were clamoring for them all the time, for years, included the Annotated Declarative system, the "batch insertmanyvalues" feature, and transparent query caching.  These were monumental tasks to implement but virtually everyone wanted these, so we did them (took years).   

For "I want to refresh this specific set of related collections on these already-loaded parent objects without querying for them and I dont want to use the wiki recipe", this is not too unreasonable of a feature but the API for this would have to be very well thought out (i.e. not the way we have "populate_existing" and "refresh" at the same time, both doing the "same" thing but differently, with different limitations on both...that's wasted complexity and effort) and I fear it would be very tough to come up with this based on a single user request; usually we like to have dozens of users asking for similar things for a long time so we can work out exactly what we should do that accommodates as many use cases as possible in a consistent way.

put another way I dont want to just "bolt on" a "many" version of refresh(), that would need a lot of refactoring to get there, and at the end of the day there was a much better and more succinct way to do this with some kind of new construct that is more tailored to what people actually need and more flexible.    

if you are omitting the primary query and you want to batch load for relationships, that's a very specific need and it seems  a lot more like a function that would work against *one relationship at a time* given a list of objects to populate.   that is, it's the recipe you have, just with some helpers.  A helper to write the query using the selectinload strategy's internals?  that can be useful.  But let's not weld it to that, because people definitely want to populate collections with *other* sets of objects too and will want to write their own queries.   so another helper to take a list of "related objects" and knit them onto "primary objects"?   That seems like what would be most flexible here.
